### DISCUSSION TITLE
集約よりも結合を先に使ったほうがいい理由

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#65] [General]
## 理由
集約→結合の順に実行されるSQLを書くと、集約したテーブルを保持するために中間テーブル(派生テーブル)を作成することになることが多い
中間テーブルを作成することで以下2点がパフォーマンスに悪影響を及ぼす可能性が高いため避けたほうがいい
- DBのメモリを圧迫する
- 既存のindexが効かず、中間テーブル用の<auto_key0>という一時indexを新規作成する

## 具体例
itemsとsale_historiesという2つのテーブルがある。item毎にどれだけ売れたかの表がほしいと仮定する

**itemsテーブル**
| item_no | item |
| --- | --- |
| 10 | FD |
| 20 | CD-R |
| 30 | MO |
| 40 | DVD |

**sale_historiesテーブル**
| sale_date | item_no | quantity |
| --- | --- | --- |
| 2007-10-01 | 10 | 4 |
| 2007-10-01 | 20 | 10 |
| 2007-10-01 | 30 | 3 |
| 2007-10-03 | 10 | 32 |
| 2007-10-03 | 30 | 12 |
| 2007-10-04 | 20 | 22 |
| 2007-10-04 | 30 | 7 |

**ほしい結果**
| item_no | total_qty |
| --- | --- |
| 10 | 36 |
| 20 | 32 |
| 30 | 22 |
| 40 | |

## 集約→結合の順に実行されるSQL
以下のようなコードを書くと集約→結合の順に実行される(サブクエリが発行された後、JOINが実行されるため)
```sql
SELECT IT.item, SH.count
FROM items IT
LEFT OUTER JOIN (select item_id, SUM(quantity) AS count FROM `sales_histories` GROUP BY item_id) SH
ON IT.ID = SH.item_id;
```
このクエリだとサブクエリで集約したテーブルを中間テーブルとしてメモリにキャッシュしてしまう(DERIVEDがサブクエリや一時的なテーブルから派生した中間テーブル)。加えて、中間テーブル作成すると同時に、結合時の参照が楽になるよう[<auto_key0>](https://gihyo.jp/article/2023/08/mysql-rcn0203)という一時的なindexを作成する。
1. itemsテーブルをフルスキャン
2. 中間テーブルを作成
3. 中間テーブルからitem_noを参照し、結合(derived2は、おそらくid=2のDERIVEDを指していると思われる)

| id	| select_type	| table	| partitions	| type	| possible_keys	| key	 | key_len	| ref	| rows	| filtered	| Extra |
| ---	| ---	| ---	| ---	| ---	| ---	| ---	 | ---	|---	| ---	| ---	| --- |
| 1	| PRIMARY	| IT	| NULL	| ALL	| NULL	| NULL	| NULL	| NULL	| 4	| 100.00	| NULL |
| 1	| PRIMARY	| \<derived2\>	| NULL	| ref	| <auto_key0>	| <auto_key0>	| 5	| mysql_sandbox.IT.id	| 2	| 100.00	| NULL |
| 2	| DERIVED	| sales_histories	| NULL	| index	| idx_item_id	| idx_item_id	| 5	| NULL	| 7	| 100.00	| NULL |

## 結合→集約の順に実行されるSQL
以下のようなコードを書くと結合→集約の順に実行される
```sql
SELECT IT.item, SUM(SH.quantity) as count
FROM items IT
LEFT OUTER JOIN sales_histories SH 
ON IT.id = SH.item_id
GROUP BY IT.id;
```
このクエリだと中間テーブルを作成しない
1. itemsテーブルをindexスキャン
3. sale_historiesテーブルからitem_noを参照し、結合

| id	| select_type	| table	| partitions	| type	| possible_keys	| key	 | key_len	| ref	| rows	| filtered	| Extra |
| ---	| ---	| ---	| ---	| ---	| ---	| ---	 | ---	|---	| ---	| ---	| --- |
| 1	| SIMPLE	| IT	| NULL	| index	| PRIMARY	| PRIMARY	| 4	| NULL	| 4	| 100.00	| NULL |
| 1	| SIMPLE	| SH	| NULL	| ref	| idx_item_id	| idx_item_id	| 5	| mysql_sandbox.IT.id	| 2	| 100.00	| NULL |

## 結論
- FROM, ON , JOINにサブクエリを使って集約を使った中間テーブルを作成していたら、先に結合することで解決できないか考える
  - https://speakerdeck.com/soudai/pgcon21j-tutorial?slide=15
- 結合は１対1でなくとも１対多ならば行数は増えない
  - Nをitemsテーブルのレコード数、Mをsales_historiesのレコード数と仮定すると、[クロス結合](https://qiita.com/devopsCoordinator/items/3ba154429973a4ddc9e0)ではレコード総数はN×Mになる
  - 1対多でLEFT OUTER JOINをする分には、クロス結合は起きず、結合後のレコード数が右側のレコード数Mを超えることはないことから、行数は増えないと言っているのだと思う

**Comment by yuyakawai0411** [comment]
## 思いつき
SQLは集合指向なのであれば、以下のベン図集をSQL文に変換することができれば、SQLを書くのが楽になりそう(A∋Bのときは、SQLでこう書くなど)
JOINを使って書く場合と集合演算子を使って書く2パターンがあると思う
https://www.u.tsukuba.ac.jp/~hashimoto.kouji.fu/paper/set_operations.html

**Comment by kakeru-one** [comment]
## 前半のまとめ
以下の点から、`GROUP BY`等の集約よりも結合の方を先に行った方が良い。
- 中間ビューを保存しないといけなくなるので、メモリ効率が悪い。
- ビューには主キーのインデックスは引き継がれないので、読み取り速度が遅くなる。（フルスキャン）

## 後半のまとめ
- 集合演算に外部結合が使えるよねって話。

**Reply by kakeru-one** [reply]
1対多の場合に使える手法である。
多対多の時は、片方を集約して1にする。

**Comment by kakeru-one** [comment]
## EXPLAIN ANALYZEとEXPLAINの出力の比較
1. `EXPLAIN ANALYZE`の`Table scan on IT  (cost=0.35 rows=1) (actual time=0.0167..0.0167 rows=0 loops=1)`は、EXPLAINの`id=1, table=IT`と一致する。
2. `EXPLAIN ANALYZE`の`Index lookup on SH using <auto_key0> (item_no=IT.item_no)  (cost=0.35..0.7 rows=2) (never executed)`は、EXPLAINの`id=1, table=<derived2>`と一致する。
3. `EXPLAIN ANALYZE`の`Materialize  (cost=0..0 rows=0) (never executed)`は、EXPLAINの`id=2, select_type=DERIVED`と一致する。

### EXPLAIN ANALYZE
```bash
| -> Nested loop left join  (cost=0.85 rows=0) (actual time=0.03..0.03 rows=0 loops=1)
    -> Table scan on IT  (cost=0.35 rows=1) (actual time=0.0167..0.0167 rows=0 loops=1)
    -> Index lookup on SH using <auto_key0> (item_no=IT.item_no)  (cost=0.35..0.7 rows=2) (never executed)
        -> Materialize  (cost=0..0 rows=0) (never executed)
            -> Table scan on <temporary>  (never executed)
                -> Aggregate using temporary table  (never executed)
                    -> Table scan on sales_history  (cost=0.35 rows=1) (never executed)
 |
```

### EXPLAIN
![スクリーンショット 2024-07-27 14 20 27](https://github.com/user-attachments/assets/fcda100d-4843-41d9-bc61-cc1ff0472735)

**Reply by kakeru-one** [reply]
EXPLAIN ANALYZEの仕様については、以下のMySQLのissueに書いてありそう。
https://dev.mysql.com/worklog/task/?id=4168

**Reply by kakeru-one** [reply]
もっとフランクなやつはこれ
https://www.percona.com/blog/using-explain-analyze-in-mysql-8/#:~:text=Explain%20Analyze.-,Explain%20Analyze,-MySQL

**Reply by kakeru-one** [reply]
`<derived2>`とDERIVEDは関連があるかも？

**Comment by kakeru-one** [comment]
P166のCROSS JOINの話。
- CROSS JOINは全ての組み合わせを出す。
- その点で、GROUP BYと同じような考え方。

DATAビューとMASTERビューは、(age_class, sex_cd)の全ての組み合わせに対して表を生成している。

具体的にいうと、
MASTERビューでは`TblAge CROSS JOIN TblSex`で(age_class, sex_cd)の全ての組み合わせを生成している。
DATAビューは、`GROUP BY age_class, sex_cd`で(age_class, sex_cd)の全ての組み合わせの集約データを生成している。

結果として、DATAビューとMASTERビューは、
(age_class, sex_cd)の組み合わせの単位で一致しているので、
(age_class, sex_cd)の組み合わせの一致をON句の条件に指定することで、`LEFT OUTER JOIN`でうまく結合できる。
